worker/src/services/ClickhouseWriter/index.ts

```diff
@@ -171,32 +172,53 @@ export class ClickhouseWriter {
       logger.error(`ClickhouseWriter.flush ${tableName}`, err);
 
       // Re-add the records to the queue with incremented attempts
-      queueItems.forEach((item) => {
+      for (const item of queueItems) {
         if (item.attempts < this.maxAttempts) {
           entityQueue.push({
             ...item,
             attempts: item.attempts + 1,
           });
         } else {
-          // TODO - Add to a dead letter queue in Redis rather than dropping
+          const shardNames = IngestionQueue.getShardNames();
+          let job = null;
+
+          for (const shardName of shardNames) {
+            const queue = IngestionQueue.getInstance({ shardName });
+            if (queue) {
+              job = await queue.getJob(item.jobId);
+              if (job) {
+                break;
+              }
+            }
+          }
+          if (job) {
+            await job.moveToFailed(
+              new Error(
+                `Max attempts reached for ${tableName} record. Dropping record.`,
+              ),
+              "token",
+            );
+          }
```

Comment:
My current understanding of our logic is that the job will be acknowledged and removed from the queue as soon as it's written into the ClickHouse Writer. Therefore, my expectation would be that `job` should never be found in any of the shards which makes this change a no-op.

Instead, we could either insert a new job into the IngestionQueue to keep cycling on the event or add it to some new "failed"/DLQ that won't be processed automatically.

Have you tried this new behaviour and found it working for you? If yes, please let me know if I miss anything here.